src/ObjectIndex.cc

```diff
@@ -0,0 +1,173 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+// Class header
+#include "lsst/partition/ObjectIndex.h"
+
+// System headers
+#include <stdexcept>
+#include <vector>
+
+// LSST headers
+#include "lsst/partition/Chunker.h"
+#include "lsst/partition/Constants.h"
+
+using namespace std;
+
+namespace lsst {
+namespace partition {
+
+ObjectIndex::~ObjectIndex() {
+    close();
+}
+
+
+void ObjectIndex::create(string const& fileName,
+                         csv::Editor const& editor,
+                         string const& idFieldName,
+                         string const& chunkIdFieldName,
+                         string const& subChunkIdFieldName) {
+    string const context = "ObjectIndex::" + string(__func__) + ": ";
+    lock_guard<mutex> lock(_mtx);
+    if (_isOpen) return;
+    if (fileName.empty()) {
+        throw invalid_argument(context + "file name is empty");
+    }
+    if (idFieldName.empty() || chunkIdFieldName.empty() || subChunkIdFieldName.empty()) {
+        throw invalid_argument(
+            context + "at least one of the required field names isn't provided,"
+            " idFieldName='" + idFieldName + "', chunkIdFieldName='" + chunkIdFieldName + "',"
+            " subChunkIdFieldName='" + subChunkIdFieldName + "'");
+    }
+    _outFileName = fileName;
+    _outFile.open(_outFileName, ios_base::out|ios_base::app);
+    if (not _outFile.good()) {
+        throw runtime_error(context + "failed to open/create index file: '" + _outFileName +  "'");
+    }
+
+    // Initialize the editor which will be used for formatting text writtent into
+    // the output file. Initialize indexes of the fields to optimize further operations
+    // with the enditor.
+    vector<string> const fields = {idFieldName, chunkIdFieldName, subChunkIdFieldName};
+    _outEditorPtr.reset(new csv::Editor(editor.getOutputDialect(), editor.getOutputDialect(), fields, fields));
+    _outIdField = _outEditorPtr->getFieldIndex(idFieldName);
+    _outChunkIdField = _outEditorPtr->getFieldIndex(chunkIdFieldName);
+    _outSubChunkIdField = _outEditorPtr->getFieldIndex(subChunkIdFieldName);
+
+    // Mallocate buffer to be large nough to accomodate teh largest possible result
+    // returned by csv::Editor::writeRecord.
+    _outBuf.reset(new char[MAX_LINE_SIZE + 1]);
+
+    _mode = Mode::WRITE;
+    _isOpen = true;
+}
+
+
+void ObjectIndex::open(string const& url,
+                       csv::Dialect const& dialect) {
+    string const context = "ObjectIndex::" + string(__func__) + ": ";
+    lock_guard<mutex> lock(_mtx);
+    if (_isOpen) return;
```

Comment:
This brings us back to the over-complicated design of the "map-reduce" machinery which constructs one `Worker` per thread and prevents us from directly passing an instance of the `ObjectIndex` into each worker. As a result of this each worker wants to create or open the index using the same set of parameters. Throwing an exception when constructing the second such `Worker` would kill the process right away.
Other options would would result in even more complicated design of this "map-reduce" machinery.
That's one of the reasons why I don't like the current implementation of the partitioner.

Reply:
Can you create that instance once in the main thread instead of forcing every worker to repeat it? I'm not sure what exactly prevents you from passing `ObjectIndex` instance to workers via parameters instead using singleton?

Reply:
Just have a deeper look at the design of the "map-reduce" to see that it doesn't allow passing any parameters to constructors of "worker" classes used in instantiating the "jobs". The idea was that those classes don't need anything but `boost::program_options::variables_map`.

Reply:
You can still "pass" it as a static (pointer) member of Worker class, this should be enough for you case as they are really sharing the same instance.

Reply:
It's "catch 22". You can't do this before `Worker` injects its group of `part.*` parameters into Google options, and before these options get parsed by the command-line parser. This makes its impossible to open/create the index in `main`.